src/listen/ldap_sync/active_directory.c

```diff
@@ -0,0 +1,684 @@
+
+	while (fr_sbuff_remaining(sbuff)) {
+		fr_sbuff_adv_until(sbuff, SIZE_MAX, &filter_terms, '\0');
+		switch (fr_sbuff_current(sbuff)[0]) {
+		case ':':
+		{
+			fr_sbuff_marker_t	ext_marker;
+
+			fr_sbuff_marker(&attr_e_marker, sbuff);
+			fr_sbuff_advance(sbuff, 1);
+			fr_sbuff_marker(&ext_marker, sbuff);
+			if ((fr_sbuff_adv_until(sbuff, SIZE_MAX, &filter_terms, '\0') == 0) ||
+			    (fr_sbuff_current(sbuff)[0] != ':')) {
+				fr_strerror_const("Extensible match filters need two \":\" surrounding the rule OID");
+				return -1;
+			}
+			if (strncmp(fr_sbuff_current(&ext_marker), LDAP_MATCHING_RULE_BIT_AND,
+				    fr_sbuff_diff(sbuff, &ext_marker)) == 0) {
+				node->op = LDAP_FILTER_OP_BIT_AND;
+				break;
+			}
+			if (strncmp(fr_sbuff_current(&ext_marker), LDAP_MATCHING_RULE_BIT_OR,
+				    fr_sbuff_diff(sbuff, &ext_marker)) == 0) {
+				node->op = LDAP_FILTER_OP_BIT_OR;
+				break;
+			}
+			fr_strerror_const("Unsupported extensible match filter");
+			return -1;
+		}
+
+		case '=':
+			if (fr_sbuff_remaining(sbuff) < 1) {
+				fr_strerror_const("Missing value after \"=\"");
+				return -1;
+			}
+
+			/*
+			 *	If we've not already set where the end of the attribute is, set it now
+			 */
+			if (fr_sbuff_diff(&attr_s_marker, &attr_e_marker) == 0) fr_sbuff_marker(&attr_e_marker, sbuff);
+
+			node->attr = talloc_zero_array(node, char, fr_sbuff_diff(&attr_e_marker, &attr_s_marker) + 1);
+			memcpy(node->attr, fr_sbuff_current(&attr_s_marker),
+			       fr_sbuff_diff(&attr_e_marker, &attr_s_marker));
+
+			if (node->op == LDAP_FILTER_OP_UNSET) node->op = LDAP_FILTER_OP_EQ;
+
+			fr_sbuff_marker(&value_marker, sbuff);
+			fr_sbuff_advance(&value_marker, 1);
+
+			/*
+			 *	Check for the attribute needed for the filter in
+			 *	the query attribute list.
+			 */
+			ldap_query_attrs_check(attrs, node->attr, "used in filter");
+
+			break;
+
+		case ')':
+			/*
+			 *	"present" filter is a special case where the value is not needed
+			 */
+			if ((fr_sbuff_diff(sbuff, &value_marker) == 1) &&
+			    (fr_sbuff_current(&value_marker)[0] == '*') &&
+			    (node->op == LDAP_FILTER_OP_EQ)) {
+				node->op = LDAP_FILTER_OP_PRESENT;
+				break;
+			}
+
+			node->value = fr_value_box_alloc_null(node);
+
+			switch (node->op) {
+			case LDAP_FILTER_OP_EQ:
+				fr_value_box_bstrndup(node, node->value, NULL, fr_sbuff_current(&value_marker),
+						      fr_sbuff_diff(sbuff, &value_marker), false);
+
+				/*
+				 *	Filter values with * in them are substring matches rather than equality
+				 */
+				if (strchr(node->value->vb_strvalue, '*')) node->op = LDAP_FILTER_OP_SUBSTR;
+				break;
+
+			/*
+			 *	Since we don't have the LDAP schema, we make an assumption that <=, >= and
+			 *	bitwise operators are going to be used with numeric attributes
+			 */
+			case LDAP_FILTER_OP_GE:
+			case LDAP_FILTER_OP_LE:
+			case LDAP_FILTER_OP_BIT_AND:
+			case LDAP_FILTER_OP_BIT_OR:
+				fr_value_box_from_str(node, node->value, FR_TYPE_UINT32, NULL,
```

Comment:
Check error condition here.  There's tons of reasons the parsing operations can fail and we definitely want to report them.